### PR TITLE
[doc] Add transfer learning E2E tutorial

### PR DESCRIPTION
**Pull request by mbencer**
This commit adds E2E (preparing dataset and model, training, inference) tutorial.

ONE-DCO-1.0-Signed-off-by: Mateusz Bencer <m.bencer@partner.samsung.com>

### CODE REVIEW

File: docs/runtime/transfer_learning.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+                                               weights='imagenet')
+
+base_model.trainable = False
+image_batch = tf.zeros([32, 160, 160, 3])
+feature_batch = base_model(image_batch)
+
+# feature extraction
+global_average_layer = tf.keras.layers.GlobalAveragePooling2D()
+feature_batch_average = global_average_layer(feature_batch)
+prediction_layer = tf.keras.layers.Dense(1, activation='relu')
```
Comment (mbencer) [inline docs/runtime/transfer_learning.md:67]: The better choice here seems to be `sigmoid` + `categorical crossentropy` as a loss function, but `sigmoid` is not supported yet.

File: docs/runtime/transfer_learning.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+In this tutorial the inference will be launched via `Python API`.
+### Convert the trained model to nnpackage format
+The current version of `Python API` expect models in Python nnpackge format. The conversion can be done with `tools/nnpackage_tool/model2nnpkg/model2nnpkg.py` script.
+```bash
+python model2nnpkg.py -m customized_mobilenetv2_trained.circle
+```
+### Run inference via Python API
+PRE: Install nnfwapi package
+```
+pip install -i https://test.pypi.org/simple/ nnfwapi==0.1.1
```
Comment (mbencer) [inline docs/runtime/transfer_learning.md:153]: I have a problem to build Python API from scratch, so I've finally used version from pypi. The problems are described in https://github.com/Samsung/ONE/issues/13332.

File: docs/runtime/transfer_learning.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+nnpackage_path = '/home/m.bencer/workspace/OneLearningNotes/test_models/mobilenetv2/customized_mobilenetv2_trained'
+session = nnfw_session(nnpackage_path, 'cpu')
+
+_URL = 'https://storage.googleapis.com/mledu-datasets/cats_and_dogs_filtered.zip'
+path_to_zip = tf.keras.utils.get_file('cats_and_dogs.zip', origin=_URL, extract=True)
+PATH = os.path.join(os.path.dirname(path_to_zip), 'cats_and_dogs_filtered')
+
+validation_dir = os.path.join(PATH, 'validation')
+
+BATCH_SIZE = 1 # we take only a single image per inference
```
Comment (mbencer) [inline docs/runtime/transfer_learning.md:171]: At first I've planed to inference more images (the model has dynamic batch) but [set_input_tensorinfo](https://github.com/Samsung/ONE/blob/1389776f542af795e13ad9efe03393fc252dbe69/runtime/onert/api/python/include/nnfw_api_wrapper.h#L123) causes segfault probably because it should be called before `nnfw_prepare` which is already called in [NNFW_SESSION ctor](https://github.com/Samsung/ONE/blob/1389776f542af795e13ad9efe03393fc252dbe69/runtime/onert/api/python/src/nnfw_api_wrapper.cc#L164)

File: docs/runtime/transfer_learning.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
+# Transfer Learning
+
+Let's clasify dogs and cats using the pre-trained model. It will be `mobilenet v2` in our case.
```
Comment (glistening) [inline docs/runtime/transfer_learning.md:3]:
```suggestion
Let's classify dogs and cats using the pre-trained model. It will be `mobilenet v2` in our case.
```

**Review by ragmani — DISMISSED** [review]
LGTM

**Comment by zetwhite** [comment]
Sorry for the late review. I'm reading it now.

File: docs/runtime/transfer_learning.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+                                               include_top=False,
+                                               weights='imagenet')
+
+base_model.trainable = False
+image_batch = tf.zeros([32, 160, 160, 3])
+feature_batch = base_model(image_batch)
+
+# feature extraction
+global_average_layer = tf.keras.layers.GlobalAveragePooling2D()
+feature_batch_average = global_average_layer(feature_batch)
```
Comment (zetwhite) [inline docs/runtime/transfer_learning.md:66]: How about removing this line?
Since `feature_batch_average` is not used in this script, So when I first read the script, I wondered why this line was included. 

```suggestion
```

File: docs/runtime/transfer_learning.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+base_model = tf.keras.applications.MobileNetV2(input_shape=IMG_SHAPE,
+                                               include_top=False,
+                                               weights='imagenet')
+
+base_model.trainable = False
+image_batch = tf.zeros([32, 160, 160, 3])
+feature_batch = base_model(image_batch)
+
+# feature extraction
```
Comment (zetwhite) [inline docs/runtime/transfer_learning.md:64]: How about this comment? 

For me, It was hard to catch the meaning of `feature extraction`.
So I checked the reference link (you linked under) - it said base_model do feature extraction and additional layer do classification. So I suggest under comments.  

>  By specifying the include_top=False argument, you load a network that doesn't include the classification layers at the top, which is ideal for feature extraction. 
( https://www.tensorflow.org/tutorials/images/transfer_learning#:~:text=By%20specifying%20the%20include_top%3DFalse%20argument%2C%20you%20load%20a%20network%20that%20doesn%27t%20include%20the%20classification%20layers%20at%20the%20top%2C%20which%20is%20ideal%20for%20feature%20extraction.) 

```suggestion
# define classification layer 
```

```suggestion
# define additional layer 
```

File: docs/runtime/transfer_learning.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+## Convert tflite model to circle model
+In the moment of writing, the `circle+` format is developed so it will the other possibility to handle model training.
+Note that `one-import` can be found in the installaion directory (determined by `-DCMAKE_INSTALL_PREFIX` cmake flag which can be added to `./nnfw configure` command).
+```bash
+one-import tflite -i customized_mobilenetv2.tflite -o customized_mobilenetv2.circle
+```
+
+## Run training
+Let's train the model using `onert_train`. In this case we want to test only 2 last layers: `Mean` and `FullyConnected` (created from `GlobalAveragePooling2D` and `Dense` in TensorFlow) Note that the other possibility is using `circle+` capabilities.
+NOTE TO REMOVE: Due to problems with exporing, the training was launched on https://github.com/Samsung/ONE/pull/13226
```
Comment (zetwhite) [inline docs/runtime/transfer_learning.md:100]:
```suggestion
NOTE TO REMOVE: Due to problems with exporting, the training was launched on https://github.com/Samsung/ONE/pull/13226
```

File: docs/runtime/transfer_learning.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+pip install -i https://test.pypi.org/simple/ nnfwapi==0.1.1
+```
+```python
+from nnfwapi.libnnfw_api_pybind import *
+
+import matplotlib.pyplot as plt
+import os
+import tensorflow as tf
+
+nnpackage_path = '/home/m.bencer/workspace/OneLearningNotes/test_models/mobilenetv2/customized_mobilenetv2_trained'
```
Comment (zetwhite) [inline docs/runtime/transfer_learning.md:162]: How about guide reader to update the `nnpackage_path` to their local path? 

```suggestion
nnpackage_path = '{your-working-dir}/mobilenetv2/customized_mobilenetv2_trained'
```

**Review by zetwhite — COMMENTED** [review]
I followed the tutorial and it also worked well in my local environment! :+1:

File: docs/runtime/transfer_learning.md (hunk 8 above)
Comment (mbencer) [reply docs/runtime/transfer_learning.md]: thank you for catching it
added `-o mobilenet_package_dir` for `model2nnpkg.py` call and use this name here

File: docs/runtime/transfer_learning.md (hunk 7 above)
Comment (mbencer) [reply docs/runtime/transfer_learning.md]: right, it's great that you catch it, because line line `NOTE TO REMOVE: ...` is no more valid since the problem resolved on master

File: docs/runtime/transfer_learning.md (hunk 6 above)
Comment (mbencer) [reply docs/runtime/transfer_learning.md]: right, I see, how about `define custom classification layers`?

File: docs/runtime/transfer_learning.md (hunk 5 above)
Comment (mbencer) [reply docs/runtime/transfer_learning.md]: ohh right, it's some leftover, removed ;)

File: docs/runtime/transfer_learning.md (hunk 6 above)
Comment (zetwhite) [reply docs/runtime/transfer_learning.md]: Looks great!

File: docs/runtime/transfer_learning.md [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+current_dir = pathlib.Path(__file__).parent
+with open(current_dir/'cats_and_dogs.input.bin', 'wb') as in_stream:
+    with open(current_dir/'cats_and_dogs.output.bin', 'wb') as out_stream:
+        for image,label in train_dataset:
+            in_stream.write(image.numpy().astype(np.float32).tobytes())
+            out_stream.write(label.numpy().astype(np.float32).tobytes())
+```
+
+## Feature extraction
```
Comment (zetwhite) [inline docs/runtime/transfer_learning.md:43]: It's minor things. How about revisiting this name?

In the same reason in my old comments - https://github.com/Samsung/ONE/pull/13329#discussion_r1665180014. 

In my preference, 
```suggestion
## Prepare fine-tuning model 
```

or 
```suggestion
## Prepare model for fine tuning
```

Comment (zetwhite) [reply docs/runtime/transfer_learning.md:43]: But it's just a suggestion, You could keep going with current version :smile:

**Review by zetwhite — APPROVED** [review]
LGTM :+1:

File: docs/runtime/transfer_learning.md (hunk 9 above)
Comment (mbencer) [reply docs/runtime/transfer_learning.md]: Sorry, I've missed it `Prepare model for fine-tuning` looks ok for me ;)

**Review by zetwhite — APPROVED** [review]

**Review by ragmani — APPROVED** [review]

**Review by glistening — APPROVED** [review]
LGTM
